spotbugs/src/main/java/edu/umd/cs/findbugs/detect/ConstructorThrow.java

```diff
@@ -0,0 +1,151 @@
+package edu.umd.cs.findbugs.detect;
+
+import java.util.HashSet;
+import java.util.Set;
+
+import org.apache.bcel.Const;
+
+import edu.umd.cs.findbugs.BugInstance;
+import edu.umd.cs.findbugs.BugAccumulator;
+import edu.umd.cs.findbugs.BugReporter;
+import edu.umd.cs.findbugs.bcel.OpcodeStackDetector;
+import org.apache.bcel.classfile.Method;
+import org.apache.bcel.classfile.ExceptionTable;
+import org.apache.bcel.classfile.JavaClass;
+import org.apache.bcel.classfile.Code;
+
+/**
+ * This detector can find constructors that throw exception.
+ */
+public class ConstructorThrow extends OpcodeStackDetector {
+
+    private final BugReporter bugReporter;
+    private final BugAccumulator bugAccumulator;
+    private final Set<String> calledFromCtor = new HashSet<String>();
+
+    private boolean isFinalClass = false;
+    private boolean isFinalFinalizer = false;
+    private boolean isFirstPass = true;
+
+    public ConstructorThrow(BugReporter bugReporter) {
+        this.bugReporter = bugReporter;
+        this.bugAccumulator = new BugAccumulator(bugReporter);
+    }
+
+    /**
+     * Visit a class to find the constructor, then collect all the methods that gets called in it.
+     * Also we are checking for final declaration on the class, or a final finalizer, as if present
+     * no finalizer attack can happen.
+     */
+    @Override
+    public void visit(JavaClass obj) {
+        resetState();
+        if (obj.isFinal()) {
+            isFinalClass = true;
+            return;
+        }
+        for (Method m : obj.getMethods()) {
+            // First visit the constructor, it might not be at the start of the class.
+            if ("<init>".equals(m.getName())) {
+                // This will visit all constructors.
+                doVisitMethod(m);
+            } else if ("finalize".equals(m.getName())) {
+                // Check for finalizer.
+                if (m.isFinal()) {
+                    isFinalFinalizer = true;
+                }
+            }
+        }
+        isFirstPass = false;
+    }
+
+    @Override
+    public void visit(Method obj) {
+        if (isFinalClass || isFinalFinalizer) {
+            return;
+        }
+        if (isConstructor() || methodCalledFromCtor()) {
+            // Check if there is a throws keyword for checked exceptions.
+            ExceptionTable tbl = obj.getExceptionTable();
+            boolean throwsExceptions = tbl != null && tbl.getNumberOfExceptions() > 0;
```

Comment:
Why create a local variable here to use it only once? Especially assigning a boolean, which is always messy (mixing `=` with `!=` and `==`)

Reply:
Thanks!  I will clean this part up.